packages/api/src/repository/storage/S3StorageClient.ts

```diff
@@ -0,0 +1,140 @@
+import { SignedUrlParameters, StorageClient, File } from './StorageClient'
+import {
+  GetObjectCommand,
+  GetObjectCommandOutput,
+  S3Client,
+  ListObjectsV2Command,
+  PutObjectCommand,
+} from '@aws-sdk/client-s3'
+import { getSignedUrl } from '@aws-sdk/s3-request-presigner'
+import type { Readable } from 'stream'
+
+// While this is listed as S3, for self hosting we will use MinIO, which is
```

Comment:
Can we use Cloudflare R2 as well for self hosting? What was the decision behind using MinIO?

Reply:
I'm not actually familiar with R2 - but anything that is S3 Compatible should work. Let me take a look later to see whether or not the Storage Client I built works with it. 

Minio was chosen because it can be self-hosted along with the rest of the application. There is a docker image, and it can all run on the same server without relying on anything external. 

I'm trying to ensure everything here can be run self-contained without any need for external services. 

That said, as with some of the email changes, I am looking into ways to simplify parts of it too, and having some external services is ok with me.

Reply:
To find suitable services, I recommend consulting r/self-hosted

Reply:
S3 is a nice idea, provides various options. How about local storage? This would reduce the required dependencies by one.

Reply:
oh wow I didn't know Minio can be self-hosted! That sounds like a good idea.

Reply:
> S3 is a nice idea, provides various options, including self hosted ones.
> 
> How about local storage? This would reduce the required dependencies by one.

The uploads are done via signed-urls, so while local-storage would be feasible it'd require a bit more development work.